packages/vitest-pool-workers/src/worker/index.ts

```diff
@@ -0,0 +1,228 @@
+	if (args[0] === -0.5) return;
+	return originalClearTimeout.apply(globalThis, args);
+};
+
+// Wraps a `WebSocket` with a Node `MessagePort` like interface
+class WebSocketMessagePort extends events.EventEmitter {
+	constructor(private readonly socket: WebSocket) {
+		super();
+		socket.addEventListener("message", this.#onMessage);
+		socket.accept();
+	}
+
+	#onMessage = (event: MessageEvent) => {
+		assert(typeof event.data === "string");
+		const parsed = structuredSerializableParse(event.data);
+		this.emit("message", parsed);
+	};
+
+	postMessage(data: unknown) {
+		const stringified = structuredSerializableStringify(data);
+		try {
+			this.socket.send(stringified);
+		} catch (error) {
+			__console.error("Error sending message to pool:", error);
+		}
+	}
+}
+
+interface JsonError {
+	message?: string;
+	name?: string;
+	stack?: string;
+	cause?: JsonError;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function reduceError(e: any): JsonError {
+	return {
+		name: e?.name,
+		message: e?.message ?? String(e),
+		stack: e?.stack,
+		cause: e?.cause === undefined ? undefined : reduceError(e.cause),
+	};
+}
+
+let patchedFunction = false;
+function ensurePatchedFunction(unsafeEval: UnsafeEval) {
+	if (patchedFunction) return;
+	patchedFunction = true;
+	// `new Function()` is used by `@vitest/snapshot`
+	globalThis.Function = new Proxy(globalThis.Function, {
+		construct(_target, args, _newTarget) {
+			// `new Function()` and `UnsafeEval#newFunction()` have reversed args
+			const script = args.pop();
+			return unsafeEval.newFunction(script, "anonymous", ...args);
+		},
+	});
+}
+
+// `RunnerObject` is a singleton and "colo local" ephemeral object. Refer to:
+// https://github.com/cloudflare/workerd/blob/v1.20231206.0/src/workerd/server/workerd.capnp#L529-L543
+export class RunnerObject implements DurableObject {
+	executor: VitestExecutorType | undefined;
+
+	constructor(_state: DurableObjectState, env: Env) {
+		vm._setUnsafeEval(env.__VITEST_POOL_WORKERS_UNSAFE_EVAL);
+		ensurePatchedFunction(env.__VITEST_POOL_WORKERS_UNSAFE_EVAL);
+		setEnv(env);
+	}
+
+	async handleVitestRunRequest(request: Request): Promise<Response> {
+		assert.strictEqual(request.headers.get("Upgrade"), "websocket");
+		const { 0: poolSocket, 1: poolResponseSocket } = new WebSocketPair();
+
+		const workerDataHeader = request.headers.get("MF-Vitest-Worker-Data");
+		assert(workerDataHeader !== null);
+		const wd = structuredSerializableParse(workerDataHeader);
+		assert(typeof wd === "object" && wd !== null);
+		assert("filePath" in wd && typeof wd.filePath === "string");
+		assert("name" in wd && typeof wd.name === "string");
+		assert("data" in wd && typeof wd.data === "object" && wd.data !== null);
+
+		const port = new WebSocketMessagePort(poolSocket);
+		try {
+			const module = await import(wd.filePath);
+
+			// HACK: Internally, the Vitest worker calls `startViteNode()`, which
```

Comment:
NIT the use of `Vitest worker` here confuses me. I don't think it is a CF Worker, right?